// @dayangac/fkld/lib/molecule.js
  // FKLD Step 3 — per-edge tucking-molecule metadata.

  // Each interior edge that participates in a tucking molecule (Tachi 2010)
  // carries four scalars. AKDE v1 stores a single uniform (θ, w) on the
  // state object because every molecule on a regular N-gon pyramid has the
  // same parameters; the architectural Kirigamizer needs *per-edge* values
  // because freeform meshes have molecules of different sizes at different
  // vertices.

  // The four parallel arrays live under `fkld:edges_moleculeTheta`,
  // `fkld:edges_moleculeWidth`, `fkld:edges_moleculeDepth`, and
  // `fkld:edges_dihedralTarget`. Like FOLD's `edges_assignment`, they are
  // indexed in lockstep with `edges_vertices`: entry i ↔ edges_vertices[i].

  // An edge that has no molecule (cut edges, mountain creases on closed
  // pyramids that don't tuck, etc.) stores `null` at that index. That keeps
  // the JSON shape uniform — every parallel array has the same length as
  // the primary `edges_vertices` array, so a tool can iterate edges without
  // branching on "is this array even present?".

  // Citation roadmap:
  //   • Eq (1)–(2)        Tachi 2010 (angle/vector closure at each vertex)
  //   • Eq (3)–(4)        Tachi 2010 (range + non-negativity inequalities)
  //   • Tuck depth        Tachi 2010 / Liu et al. 2019 §3.1
var isFiniteOrNull,
  hasProp = {}.hasOwnProperty;

import {
  KEYS,
  deepFreeze
} from "./spec.js";

// Field metadata — single source of truth for unit/range/citation per
// parameter. Validators and the eventual UI tooltips read from here so
// the docstring lives next to the rule it enforces.
export var MOLECULE_FIELDS = deepFreeze({
  theta: {
    key: KEYS.edges.moleculeTheta,
    units: "rad",
    range: [
      -Math.PI,
      Math.PI // Strict inequality enforced below (C3).
    ],
    description: "Edge-molecule angle θ(i,j) — angular spread of the trapezoid at the vertex side.",
    citation: "Tachi 2010 Eq (3); -π < θ(i,j) < π.",
    nullable: true
  },
  width: {
    key: KEYS.edges.moleculeWidth,
    units: "mm",
    range: [0, 2e308],
    description: "Edge-molecule chord width w(i,j) — outer chord length of the trapezoid.",
    citation: "Tachi 2010 Eq (4); w(i,j) ≥ 0.",
    nullable: true
  },
  depth: {
    key: KEYS.edges.moleculeDepth,
    units: "mm",
    range: [0, 2e308],
    description: "Tuck depth behind the adjacent face — how far the molecule hides under the surface when folded.",
    citation: "Tachi 2010 §3D tuck-depth inequality; Liu et al. 2019 Figure 7.",
    nullable: true
  },
  dihedral: {
    key: KEYS.edges.dihedralTarget,
    units: "rad",
    range: [-Math.PI, Math.PI],
    description: "Target dihedral angle in the folded state — used by the simulator to set crease rest angles.",
    citation: "AKDE bar-and-hinge solver (kirigami/sim/forces.ts).",
    nullable: true
  }
});

// Quick helper to fetch the registered key for a field. Lets validators
// write generic error messages without hard-coding the FKLD prefix.
export var moleculeKey = function(field) {
  var meta;
  meta = MOLECULE_FIELDS[field];
  if (meta == null) {
    return null;
  }
  return meta.key;
};

// Per-scalar validators. Each returns true for valid numeric values and
// also true for `null` (the documented "no molecule here" sentinel).
// Non-number / non-null entries are rejected. NaN / ±Infinity are rejected
// even though they are typeof "number" — physical parameters can't be NaN.
isFiniteOrNull = function(value, min, max) {
  if (value === null || value === void 0) {
    return true;
  }
  if (typeof value !== "number") {
    return false;
  }
  if (Number.isNaN(value) || !Number.isFinite(value)) {
    return false;
  }
  return value > min && value < max; // strict inequalities — boundary values are degenerate molecules.
};

export var isValidTheta = function(v) {
  return isFiniteOrNull(v, -Math.PI, Math.PI);
};

export var isValidWidth = function(v) {
  return v === null || v === void 0 || (typeof v === "number" && Number.isFinite(v) && v >= 0);
};

export var isValidDepth = function(v) {
  return v === null || v === void 0 || (typeof v === "number" && Number.isFinite(v) && v >= 0);
};

export var isValidDihedral = function(v) {
  return isFiniteOrNull(v, -Math.PI, Math.PI);
};

// Bundle the four arrays into the FOLD-style record the rest of FKLD
// consumes. Lengths must already match `edges_vertices.length`; this
// helper does not pad or truncate. Use when serializing AKDE's internal
// state to a FKLD file.
export var packEdgeMolecules = function({theta, width, depth, dihedral} = {}) {
  var out;
  out = {};
  if (theta != null) {
    out[KEYS.edges.moleculeTheta] = theta;
  }
  if (width != null) {
    out[KEYS.edges.moleculeWidth] = width;
  }
  if (depth != null) {
    out[KEYS.edges.moleculeDepth] = depth;
  }
  if (dihedral != null) {
    out[KEYS.edges.dihedralTarget] = dihedral;
  }
  return out;
};

// Inverse of `packEdgeMolecules` — pull the four arrays out of a FKLD
// object, returning `undefined` for any array that's absent. Use when
// parsing an inbound FKLD file. Performs *no* validation; pair with
// `validateMoleculeArrays` if the source is untrusted.
export var unpackEdgeMolecules = function(fkld) {
  if (fkld == null) {
    return {};
  }
  return {
    theta: fkld[KEYS.edges.moleculeTheta],
    width: fkld[KEYS.edges.moleculeWidth],
    depth: fkld[KEYS.edges.moleculeDepth],
    dihedral: fkld[KEYS.edges.dihedralTarget]
  };
};

// Validate the four molecule arrays against the FOLD `edges_vertices`
// array. Like `validateEdgeCutTypes` in Step 2, accumulates every error
// in a single pass so the architect sees the full picture. Returns
// `{ ok, errors: [{ field, index, message }] }`.

// Rules:
//   1. Each array, if present, has the same length as `edges_vertices`.
//   2. Each entry is either `null`/`undefined` (no molecule) or a valid
//      scalar per the per-field rules.
//   3. A non-null θ at edge i implies a non-null w at edge i (and vice
//      versa). You can't half-specify a molecule.

// Tachi's full Eq (1)-(2) closure constraint requires the vertex topology
// of the mesh, so it lives in Step 26 of `AtoK.md` (per-vertex closure
// solver) not here. This Step-3 validator only enforces the per-edge
// numeric envelope.
export var validateMoleculeArrays = function(edgesVertices, arrays) {
  var arr, entry, errors, field, i, j, k, len, n, ref, spec, thetaPresent, validators, widthPresent;
  errors = [];
  if (!Array.isArray(edgesVertices)) {
    errors.push({
      field: null,
      index: -1,
      message: "edges_vertices must be an array"
    });
    return {
      ok: false,
      errors: errors
    };
  }
  n = edgesVertices.length;
  validators = {
    theta: {
      check: isValidTheta,
      name: "theta",
      key: KEYS.edges.moleculeTheta
    },
    width: {
      check: isValidWidth,
      name: "width",
      key: KEYS.edges.moleculeWidth
    },
    depth: {
      check: isValidDepth,
      name: "depth",
      key: KEYS.edges.moleculeDepth
    },
    dihedral: {
      check: isValidDihedral,
      name: "dihedral",
      key: KEYS.edges.dihedralTarget
    }
  };
  for (field in validators) {
    if (!hasProp.call(validators, field)) continue;
    spec = validators[field];
    arr = arrays != null ? arrays[field] : void 0;
    if (arr == null) {
      continue;
    }
    if (!Array.isArray(arr)) {
      errors.push({
        field: field,
        index: -1
      }, {
        message: `${spec.key} must be an array (got ${typeof arr})`
      });
      continue;
    }
    if (arr.length !== n) {
      errors.push({
        field: field,
        index: -1
      }, {
        message: `${spec.key} length (${arr.length}) must equal edges_vertices length (${n})`
      });
    }
    for (i = j = 0, len = arr.length; j < len; i = ++j) {
      entry = arr[i];
      if (!spec.check(entry)) {
        errors.push({
          field: field,
          index: i
        }, {
          message: `edge ${i}: ${spec.key} entry ${JSON.stringify(entry)} is invalid`
        });
      }
    }
  }
  // Pairing check: θ and w must be jointly present (or jointly absent).
  if (((arrays != null ? arrays.theta : void 0) != null) && ((arrays != null ? arrays.width : void 0) != null) && Array.isArray(arrays.theta) && Array.isArray(arrays.width) && arrays.theta.length === arrays.width.length) {
    for (i = k = 0, ref = arrays.theta.length; (0 <= ref ? k < ref : k > ref); i = 0 <= ref ? ++k : --k) {
      thetaPresent = (arrays.theta[i] != null) && arrays.theta[i] !== null;
      widthPresent = (arrays.width[i] != null) && arrays.width[i] !== null;
      if (thetaPresent !== widthPresent) {
        errors.push({
          field: "pair",
          index: i
        }, {
          message: `edge ${i}: molecule theta and width must both be specified (or both null)`
        });
      }
    }
  }
  return {
    ok: errors.length === 0,
    errors: errors
  };
};
